modules/rosapps/applications/devutils/shlextdbg/shlextdbg.cpp

```diff
@@ -7,27 +7,380 @@
+{
+    WCHAR buf[39];
+    if (String[0] == '{')
+    {
+        lstrcpynW(buf, String, 39);
+        if (SUCCEEDED(CLSIDFromString(buf, &Clsid)))
+        {
+            String = String + 38;
+            return true;
+        }
+    }
+    return false;
+}
+
+static HRESULT GetUIObjectOfAbsolute(LPCITEMIDLIST pidl, REFIID riid, void**ppv)
+{
+    CComPtr<IShellFolder> shellFolder;
+    PCUITEMID_CHILD child;
+    HRESULT hr = SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &shellFolder), &child);
+    if (SUCCEEDED(hr)) hr = shellFolder->GetUIObjectOf(NULL, 1, &child, riid, NULL, ppv);
+    return hr;
+}
+
+static HRESULT CreateShellItemFromParse(LPCWSTR Path, IShellItem**ppSI)
+{
+    PIDLIST_ABSOLUTE pidl = NULL;
+    HRESULT hr = SHParseDisplayName(Path, NULL, &pidl, 0, NULL);
+    if (SUCCEEDED(hr))
+    {
+        hr = SHCreateShellItem(NULL, NULL, pidl, ppSI);
+        SHFree(pidl);
+    }
+    return hr;
+}
+
+static void GetAssocClass(LPCWSTR Path, LPCITEMIDLIST pidl, HKEY &hKey)
+{
+    hKey = NULL;
+    IQueryAssociations *pQA;
+    if (SUCCEEDED(GetUIObjectOfAbsolute(pidl, IID_PPV_ARG(IQueryAssociations, &pQA))))
+    {
+        pQA->GetKey(0, ASSOCKEY_CLASS, NULL, &hKey); // Not implemented in ROS
+        pQA->Release();
+    }
+    if (!hKey)
+    {
+        DWORD cb;
+        WCHAR buf[MAX_PATH], *ext = PathFindExtensionW(Path);
+        SHFILEINFOW info;
+        info.dwAttributes = 0;
+        SHGetFileInfoW((LPWSTR)pidl, 0, &info, sizeof(info), SHGFI_PIDL | SHGFI_ATTRIBUTES);
+        if (info.dwAttributes & SFGAO_FOLDER)
+        {
+            ext = const_cast<LPWSTR>(L"Directory");
+        }
+        else if (info.dwAttributes & SFGAO_BROWSABLE)
+        {
+            ext = const_cast<LPWSTR>(L"Folder"); // Best guess
+        }
+        else
+        {
+            cb = sizeof(buf);
+            if (!SHGetValueW(HKEY_CLASSES_ROOT, ext, NULL, NULL, buf, &cb))
+            {
+                RegOpenKeyExW(HKEY_CLASSES_ROOT, buf, 0, KEY_READ, &hKey);
+            }
+        }
+        if (!hKey)
+        {
+            RegOpenKeyExW(HKEY_CLASSES_ROOT, ext, 0, KEY_READ, &hKey);
+        }
+    }
+}
+
+static void DumpBytes(const void *Data, SIZE_T cb)
+{
+    for (SIZE_T i = 0; i < cb; ++i)
+    {
+        wprintf(L"%s%.2X", i ? L" " : L"", ((LPCBYTE)Data)[i]);
+    }
+    wprintf(L"\n");
+}
+
+static HRESULT GetCommandString(IContextMenu&CM, UINT Id, UINT Type, LPWSTR buf, UINT cchMax)
+{
+    *buf = UNICODE_NULL;
+    HRESULT hr = CM.GetCommandString(Id, Type | GCS_UNICODE, 0, (char*)buf, cchMax);
+    if (FAILED(hr))
+    {
+        STRRET sr;
+        sr.uType = STRRET_CSTR;
+        if (SUCCEEDED(hr = CM.GetCommandString(Id, Type & ~GCS_UNICODE, 0, sr.cStr, 260)))
+            hr = StrRetToBufW(&sr, NULL, buf, cchMax);
+    }
+    return hr;
+}
+
+static void DumpMenu(HMENU hMenu, UINT IdOffset, IContextMenu*pCM, BOOL FakeInit, UINT Indent)
+{
+    WCHAR buf[MAX_PATH], recurse = Indent != UINT(-1);
```

Comment:
What's the type of "recurse" ?

Reply:
Then use BOOL (or BOOLEAN, or `bool` since this is C++ code, but depends where the recurse` is used), so that the purpose of the variable becomes clear.